rfc/plugins.md

```diff
@@ -0,0 +1,248 @@
+
+The goal of this proposal is to offer the best of both worlds by allowing extension point developers (Core Spinnaker developers) to decide what runtime(s) are the best for the extension.
+
+Extension points will be defined and exposed via PF4J.
+
+### PF4J as a Foundation
+[PF4J](https://github.com/pf4j/pf4j) is to be used as the foundation for all plugin functionality. While designed for in-process extensibility, we’ll be using PF4J to define both the contract for in-process plugins, as well as remote plugins. Remote plugins would be supported in two parts: First, by including a plugin that activates a new plugin runtime (e.g. gRPC, or a Docker co-process). Second, a remote plugin stub that uses the runtime plugin as a transport.
+
+PF4J has a lot to offer out of the box that we would need to build ourselves, regardless of final implementation. It doesn’t go as far as OSGI, but provides basic (yet extensible) ClassLoader isolation and a simple extension contract that we can use broadly. PF4J was originally built for performing in-process plugins only, but its own extension points make it feasible to enable remote plugins using the same contract. What runtime a particular plugin uses would be determined by a combination of: 1) The extension point developer allowing remote invocation, and 2) an operator selecting the particular runtime they want to use from the available options.
+
+An extension looks like the following:
+
+```kotlin
+// Definition of the PluginStage extension point.
+// The annotation defined here also informs the Spinnaker
+// plugin framwework which runtimes are acceptable. This 
+// annotation would additionally be used for other metadata
+// Spinnaker needs to know atop whatever PF4J offers out of
+// the box.
+@SpinnakerExtension(drivers = [
+  ExtensionDriver.LOCAL, 
+	ExtensionDriver.REMOTE
+])
+class PipelineStageExtension : ExtensionPoint {
+	// ...
+}
+```
+
+It is highly recommended to read the documentation of PF4J to fully understand the scope of this proposal.
+
+### Front50 as the Source of Truth
+Operators will need the ability to understand, at a glance, what plugins are installed and enabled or disabled, their versions, as well as any other metadata about a plugin. Whether plugins are implemented as in-process or not, they Front50 should be used to track and expose a single source of truth of what plugins are installed in any particular service, as well as which ones are enabled or disabled. 
+
+The Front50 plugin registry will also be exposed through Gate, and can be consumed by Deck to load the correct plugins on application startup.
+
+Plugins can also be enabled or disabled at runtime without redeploying a service. This flag is sourced from Front50, which all services will be listening for change events from. Whether this change event is published or polled for initially is unknown.
+
```

Comment:
Plugin management can be a component with in Deck that will allow operators to turn on / off .. may be ?

Reply:
Yes. I think we'll definitely want an operator view in Deck that would allow operators to see what plugins are available, installed, enabled/disabled and be able to actuate the ones that are installed (enable/disable as needed). My intention (not mentioned in this proposal) is that plugin management would be implemented as a "core" plugin to dogfood the whole process.

Reply:
Ideally (medium-/long-term), I'd like to make Fast Properties a plugin and then build the operator experience atop that.

Reply:
Remote monitors for the Monitored Deploy Strategy seem like a good candidate as well. It's unlucky timing that we already have an HTTP-based API contract defined at this point that is very specific to that use case, but thought I'd suggest it anyway.

Reply:
One more thought on plugin management: it's impossible not to think of Jenkins while reading through this, and both the flexibility that its plugin model allows and the nightmare that is managing plugins in a Jenkins installation... I wondered if we should call out whether managing plugin updates specifically are a goal or non-goal.

Reply:
> Remote monitors for the Monitored Deploy Strategy seem like a good candidate as well. It's unlucky timing that we already have an HTTP-based API contract defined at this point that is very specific to that use case, but thought I'd suggest it anyway.

P.S. Or maybe deployment strategies in general? Interested in your thoughts on where to draw the line here since you proposed stages as a candidate for plugins but I'm unclear whether it would make sense to replace an entire Deploy stage if you want to use a different deployment strategy?

Reply:
> One more thought on plugin management: it's impossible not to think of Jenkins while reading through this, and both the flexibility that its plugin model allows and the nightmare that is managing plugins in a Jenkins installation... I wondered if we should call out whether managing plugin updates specifically are a goal or non-goal.

I see the section on upgrades/downgrades down below now. I'd suggest adding that to the goals list.